File: kensa.dev/snippets/src/test/java/apidocs/FixturesSnippets.java
// Snippet source for kensa.dev/docs/api/fixtures.md — Java examples
package apidocs;

import dev.kensa.fixture.Fixture;
import dev.kensa.fixture.ParameterFixture;
import dev.kensa.fixture.Parents;
import dev.kensa.fixture.SecondaryFixture;

import java.time.LocalDate;

import static dev.kensa.fixture.FixtureKt.createFixture;
import static dev.kensa.fixture.FixtureKt.createParameterFixture;

class FixturesSnippets {

    record AppointmentSlot(LocalDate date, String slot) {}
    record ServiceAddress(String postcode, String line1, String town, String county) {}
    record TrackingId() {}

    static final Fixture<LocalDate> appointmentDate = createFixture("Appointment Date", LocalDate::now);
    static final Fixture<String> appointmentTimeSlot = createFixture("Appointment Time Slot", () -> "AM");
    static final Fixture<String> postcode = createFixture("Postcode", () -> "SW1A 1AA");
    static final Fixture<String> addressLine1 = createFixture("Address Line 1", () -> "1 Main St");
    static final Fixture<String> town = createFixture("Town", () -> "Springfield");
    static final Fixture<String> county = createFixture("County", () -> "Westshire");

    // 2 parents
    static final Fixture<AppointmentSlot> slot =
            createFixture("Appointment Slot", appointmentDate, appointmentTimeSlot,
                    (date, timeSlot) -> new AppointmentSlot(date, timeSlot));

    // more than 3 dependencies — construct a SecondaryFixture directly
    static final Fixture<ServiceAddress> serviceAddress = new SecondaryFixture<ServiceAddress>(
            "Service Address",
            fixtures -> new ServiceAddress(
                    fixtures.get(postcode), fixtures.get(addressLine1),
                    fixtures.get(town), fixtures.get(county)
            ),
            new Parents.Three<>(postcode, addressLine1, town),
            false
    );

    // highlighted
    static final Fixture<TrackingId> trackingId = createFixture("Tracking Id", /* highlighted = */ true, TrackingId::new);

    // parameter-derived
    static final ParameterFixture<String> GREETING =
            createParameterFixture("Greeting", "userName", (String name) -> "Hello, " + name);
}

File: kensa.dev/snippets/src/test/kotlin/apidocs/FixturesSnippets.kt
// Snippet source for kensa.dev/docs/api/fixtures.md — Kotlin examples
package apidocs

import com.natpryce.hamkrest.equalTo
import dev.kensa.StateCollector
import dev.kensa.WithFixturesSuite
import dev.kensa.fixture.FixtureContainer
import dev.kensa.fixture.FixtureRegistry.registerFixtures
import dev.kensa.fixture.fixture
import dev.kensa.fixture.parameterFixture
import dev.kensa.hamkrest.WithHamkrest
import dev.kensa.junit.KensaTest
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

// --- Co-locating fixtures with a test ---

class OrderPricingTest : KensaTest, WithHamkrest {

    init {
        registerFixtures(Companion)
    }

    @Test
    fun rendersFixturesDeclaredInTheCompanion() {
        then(theGreeting(), equalTo(fixtures[Greeting]))
    }

    private fun theGreeting() = StateCollector { "Hello ${fixtures[FirstName]}" }

    companion object : FixtureContainer {
        val FirstName = fixture("FirstName") { "Jane" }
        val Greeting = fixture("Greeting", FirstName) { name -> "Hello $name" }
    }
}

// --- Parameter-derived fixtures ---

object GreetingFixtures : FixtureContainer {
    val Greeting = parameterFixture("Greeting", from = "userName") { name: String ->
        "Hello, ${name.replaceFirstChar(Char::uppercase)}"
    }
}

class GreetingTest : KensaTest, WithHamkrest {

    init {
        registerFixtures(GreetingFixtures)
    }

    @ParameterizedTest
    @ValueSource(strings = ["alice", "bob"])
    fun `greets the user`(userName: String) {
        then(theBanner(), equalTo(fixtures[GreetingFixtures.Greeting]))
    }

    private fun theBanner() = StateCollector { "Hello, Alice" }
}

// --- WithFixturesSuite ---

object TelecomsFixtures : FixtureContainer {
    val AccountNumber = fixture("Account Number") { "ACC-123" }
}

interface WithTelecomsFixtures : WithFixturesSuite<TelecomsFixtures> {
    override val fixturesObject get() = TelecomsFixtures
}

class FeasibilityServiceTest : KensaTest, WithTelecomsFixtures, WithHamkrest {
    @Test
    fun `can check feasibility`() {
        then(theAccountNumber(), equalTo(fixtures { AccountNumber }))
    }

    private fun theAccountNumber() = StateCollector { fixtures { AccountNumber } }
}
